### PR TITLE
Update Sync Queue so that serialize is wrapped to catch errors

### LINKED ISSUE #7252: Sync: Fatal Error -- Serialization of Closure is not allowed
```
Fatal error: Uncaught exception 'Exception' with message 'Serialization of
'Closure' is not allowed' in wp-content/plugins/jetpack/sync/class.jetpack-sync-queue.php:53
Stack trace: #0`
`wp-content/plugins/jetpack/sync/class.jetpack-sync-queue.php(53):
serialize(Array) #1`
`wp-content/plugins/jetpack/sync/class.jetpack-sync-listener.php(218):
Jetpack_Sync_Queue->add(Array) #2`
`wp-content/plugins/jetpack/sync/class.jetpack-sync-listener.php(116):
Jetpack_Sync_Listener->enqueue_action('jetpack_sync_ca...', Array,
Object(Jetpack_Sync_Queue)) #3 [internal function]:
Jetpack_Sync_Listener->action_handler('post_types', Array) #4`
`wp-includes/class-wp-hook.php(298):
call_user_func_array(Array, Array) #5`
`wp-includes/class-wp-hook.php(323):
WP_Hook->apply_filters('', Array) #6`
`wp-includes/plugin.php(453): WP_Ho in
wp-content/plugins/jetpack/sync/class.jetpack-sync-queue.php
on line 53
```

Checking `class.jetpack-sync-queue.php` shows error at this line:

```
$rows_added = $wpdb->query( $wpdb->prepare(
'INSERT INTO $wpdb->options (option_name,
option_value, autoload) VALUES (%s, %s,%s)',
$this->get_next_data_row_option_name(),
serialize( $item ),
'no'
) );
```

Reported in 3231883-t

### PR DESCRIPTION
**Pull request by mdbitz**
Serialize can throw exceptions when called on closures. This update catches serialization errors so that the action/data fails silently allowing continued processing.

Fixes #7252 

#### Changes proposed in this Pull Request:
* Handle Fatal exception caused by anonymous functions on callables.

#### Does this pull request change what data or activity we track or use?
* NO

#### Testing instructions:

* On Master branch
* Add a hook to one of Sync's callable where the function internally uses an anonymous function. functions.https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/packages/sync/src/class-defaults.php#L278
* do_action( 'jetpack_sync_unlock_sync_callable' );
* perform an action (edit post)
* Fatal Exception :  Uncaught Exception: Serialization of 'Closure' is not allowed should be triggered
* Checkout fix/sync-serialize-closures branch
* do_action( 'jetpack_sync_unlock_sync_callable' );
* perform an action (edit post)
* Verify no fatal exceptions

#### Proposed changelog entry for your changes:
* Bug Fix : Catch serialization issues in Jetpack Sync.

### CODE REVIEW

**Comment by mdbitz** [comment]
Note :: skipped the pre-commit due to serialize warning, but later commit moved the comment appropriately.

File: projects/packages/sync/src/class-queue.php [hunk 1]
```diff
@@ -46,14 +46,22 @@ public function __construct( $id ) {
 	public function add( $item ) {
 		global $wpdb;
 		$added = false;
+
+		// Attempt to serialize data, if an exception (closures) return early.
+		try {
+			$item = serialize( $item ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
```
Comment (bisko) [inline projects/packages/sync/src/class-queue.php:52]: Why not use `maybe_serialize` and avoid the PHPCS exception?

File: projects/packages/sync/src/class-queue.php [hunk 2]
```diff
@@ -77,9 +85,14 @@ public function add_all( $items ) {
 		$rows        = array();
 		$count_items = count( $items );
 		for ( $i = 0; $i < $count_items; ++$i ) {
-			$option_name  = esc_sql( $base_option_name . '-' . $i );
-			$option_value = esc_sql( serialize( $items[ $i ] ) ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
-			$rows[]       = "('$option_name', '$option_value', 'no')";
+			try {
+				$option_name  = esc_sql( $base_option_name . '-' . $i );
+				$option_value = esc_sql( serialize( $items[ $i ] ) ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
```
Comment (bisko) [inline projects/packages/sync/src/class-queue.php:90]: Same as above comment.

**Review by bisko — CHANGES_REQUESTED** [review]

File: projects/packages/sync/src/class-queue.php (hunk 1 above)
Comment (mdbitz) [reply projects/packages/sync/src/class-queue.php:52]: I was thinking we'd hit the same error with maybe_serialize. I should make a unit test for this and test it out :)

```
function maybe_serialize( $data ) {
    if ( is_array( $data ) || is_object( $data ) ) {
        return serialize( $data );
    }
 
    /*
     * Double serialization is required for backward compatibility.
     * See https://core.trac.wordpress.org/ticket/12930
     * Also the world will end. See WP 3.6.1.
     */
    if ( is_serialized( $data, false ) ) {
        return serialize( $data );
    }
 
    return $data;
}
```

Comment (mdbitz) [reply projects/packages/sync/src/class-queue.php:52]: Verified maybe_serialize runs into the same exception.

**Comment by mdbitz** [comment]
Added Test to verify that no exception is triggered by the callable being a closure (anonymous function).

Please note that anonymous functions are considered objects and so maybe_serialize would still trigger the Exception.

**Review by jeherve — CHANGES_REQUESTED** [review]
The tests appear to be failing for this.

**Comment by mdbitz** [comment]
:sigh: I know I need to \ Exception and other classes I just never do.

Tests now pass, the All "commits were linted before commit." X is from me pushing changes prior to fixing lints. No linting issues in the final version of the code.

**Review by bisko — APPROVED** [review]

**Comment by davidwebca** [comment]
Sorry to barge in, I'm having this issue right now on wordpress.com hosting. My theme is configured with aaaaa lot of closures (as it's supposed to be supported by php / wordpress) so I'm guessing this will fix it? Any ETA? I get a blank screen of death on the widgets screen when widgets visibility option is activated 😢

**Pull request merged**